crates/pathfinder/examples/reorg_detection.rs

```diff
@@ -0,0 +1,199 @@
+        let number = block.number.unwrap();
+        let number = BlockId::Number(BlockNumber::Number(number));
+        let l1_block = w3.eth().block(number).await.unwrap();
+
+        if let Some(l1_block) = l1_block {
+            if &l1_block == block {
+                break;
+            }
+        }
+
+        delete_count += 1;
+    }
+
+    for _ in 0..delete_count {
+        history.pop_back();
+    }
+}
+
+/// Indicates that a reorg occurred.
+struct Reorg;
+
+/// Retrieve the next block from L1. If we have reached L1 HEAD, will poll until
+/// next block is available.
+///
+/// Returns [Err(Reorg)] if an L1 chain reorganization occurred.
+async fn get_next_block(w3: &Web3<Http>, local_head: &Block<H256>) -> Result<Block<H256>, Reorg> {
+    let local_number = local_head.number.expect("Block should have number");
+    let next_block = local_number + 1u64;
+    let next_block = BlockId::Number(BlockNumber::Number(next_block));
+
+    // A tracker to allow better printing.
+    let mut sleep_mode = false;
+
+    loop {
+        match w3
+            .eth()
+            .block(next_block)
+            .await
+            .expect("Failed to read block")
+        {
+            Some(block) => match block.parent_hash == local_head.hash.unwrap() {
+                true => {
+                    if sleep_mode {
+                        println!();
+                    }
+                    return Ok(block);
+                }
+                false => {
+                    if sleep_mode {
+                        println!();
+                    }
+                    return Err(Reorg);
+                }
+            },
+            None => {
+                // Block does not exist, this could be because we have reach the HEAD, or because
+                // there was a reorg and the requested block number is far ahead of reality.
+                let l1_latest = w3
+                    .eth()
+                    .block_number()
+                    .await
+                    .expect("Failed to get latest block number from Ethereum");
+
+                match l1_latest < local_number {
+                    true => {
+                        if sleep_mode {
+                            println!();
+                        }
+                        return Err(Reorg);
+                    }
+                    false => {
+                        match sleep_mode {
+                            true => {
+                                print!(".");
+                                std::io::stdout().flush().unwrap();
+                            }
+                            false => {
+                                sleep_mode = true;
+                                print!("No new block, sleeping ");
+                                std::io::stdout().flush().unwrap();
+                            }
+                        }
+
+                        tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
I was wondering where is the backoff time in the main loop so as not to spam the api, and now I've found it :smile: